packages/@react-aria/utils/src/scrollIntoView.ts

```diff
@@ -10,32 +10,83 @@
  * governing permissions and limitations under the License.
  */
 
+import {getScrollParent} from './';
+
+interface ScrollOpts {
+  /** The scroll parent of the target element.  */
+  scrollView: HTMLElement
+}
+
+interface ScrollIntoViewportOpts {
+  /** The optional containing element of the target to be centered in the viewport. */
+  containingElement?: Element
+}
+
+let isScrollPrevented = false;
+let delayScrolling = false;
+let scrollCallback: () => void;
+
+export function setScrollPrevented(value: boolean) {
+  isScrollPrevented = value;
+}
+
+export function setDelayScrolling(value: boolean) {
+  if (!value && delayScrolling && scrollCallback != null) {
+    scrollCallback();
+    scrollCallback = null;
+  }
+  delayScrolling = value;
+}
+
+/**
+ * Scrolls `opts.scrollView` so that `element` is visible.
+ * Similar to `element.scrollIntoView({block: 'nearest'})` (not supported in Edge),
+ * but doesn't affect parents above `opts.scrollView`.
+ */
+export function scrollIntoView(element: HTMLElement, opts: ScrollOpts);
 /**
  * Scrolls `scrollView` so that `element` is visible.
  * Similar to `element.scrollIntoView({block: 'nearest'})` (not supported in Edge),
  * but doesn't affect parents above `scrollView`.
+ * @deprecated - provide `scrollView` via opts instead.
  */
-export function scrollIntoView(scrollView: HTMLElement, element: HTMLElement) {
-  let offsetX = relativeOffset(scrollView, element, 'left');
-  let offsetY = relativeOffset(scrollView, element, 'top');
-  let width = element.offsetWidth;
-  let height = element.offsetHeight;
+export function scrollIntoView(scrollView: HTMLElement, element: HTMLElement);
+export function scrollIntoView(element: HTMLElement, opts: HTMLElement | ScrollOpts) {
```

Comment:
Could you explain the API change here? Is it necessary?

Reply:
nope not necessary, more along the lines of making the `scrollIntoView` and `scrollIntoViewport` signatures more alike, see https://github.com/adobe/react-spectrum/pull/3899#discussion_r1066458214 for some context.

Reply:
hmm not quite sure why we need both `scrollIntoView` and `scrollIntoViewport` as you mentioned there - doesn't `scrollIntoViewport` first scroll the container and then the element? what happens if you don't call `scrollIntoView` first?

Either way, I'd leave this function as it was (it's confusing to have two signatures that are opposites of each other), and maybe make the new one match this one, or just leave it. The options object isn't really needed if there's only one option (which is required?), unless we think there will be a lot more in the future.

Should `containerElement` in the other one even be optional? It seems to be provided in every call to the function so maybe that doesn't need to be an options object either?

Reply:
> hmm not quite sure why we need both scrollIntoView and scrollIntoViewport as you mentioned there - doesn't scrollIntoViewport first scroll the container and then the element? what happens if you don't call scrollIntoView first?

If we don't call `scrollIntoView` first and only rely on calling `scrollIntoViewport` then we run into the case where we shift the container's position needlessly when scrolling an item into view. See the below:

https://user-images.githubusercontent.com/9661127/214989391-9a4b32e5-f24f-4fe0-bf3d-37f1a6eb06b6.mov

The items being scrolled into view is "out of view" (aka not visible) and thus `targetElement?.scrollIntoView?.({block: 'nearest'});` will cause a position change. Since there was a position change, the `scrollintoViewport` [logic](https://github.com/adobe/react-spectrum/blob/0d06149c05549653cd27feba22dff6bbf4ac8cdc/packages/%40react-aria/utils/src/scrollIntoView.ts#L117-L121) will think we need to center the container. This logic is meant for a scenario like this:

https://user-images.githubusercontent.com/9661127/214991325-6e34384c-0178-4ead-adda-4b96c8faca29.mov

where we decided that it would good for the user to get the full context of where their focus moved and thus opted to scroll the container to the center of the screen. Therefore the ideal flow that I could think of was the two step process mentioned https://github.com/adobe/react-spectrum/pull/3899#discussion_r1066458214, hence the need to call `scrollIntoView` and then `scrollIntoViewport`.

> Should containerElement in the other one even be optional? It seems to be provided in every call to the function so maybe that doesn't need to be an options object either?

This was a thought for the future for when we would scroll any standalone component into view upon keyboard interaction (e.g. hitting enter on a MenuTrigger button that was scrolled off screen). I figured `options` would accept a `scrollStrategy` (i.e. 'nearest', 'center', etc) so that we could have the element be scrolled into the center of the screen instead of just the edge. In the case of these standalone components, the `containerElement` may not exist and thus is optional

> Either way, I'd leave this function as it was (it's confusing to have two signatures that are opposites of each other), and maybe make the new one match this one, or just leave it. The options object isn't really needed if there's only one option (which is required?), unless we think there will be a lot more in the future.

 I'll go ahead and revert `scrollIntoView` signature to what it used to be then. Happy to discuss `scrollIntoViewport`'s signature further